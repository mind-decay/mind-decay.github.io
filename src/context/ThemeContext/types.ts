import { Nullable } from 'types';

export type ThemeType = 'light' | 'dark';

export type ThemeContextType = {
  theme: ThemeType;
  toggleTheme: () => void;
};

export const isThemeType = (theme: Nullable<string>): theme is ThemeType => {
  return typeof theme === 'string' && ['light', 'dark'].includes(theme);
};
