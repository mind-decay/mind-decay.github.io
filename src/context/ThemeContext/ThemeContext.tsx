import { ReactNode, createContext, useContext, useEffect, useState } from 'react';
import { ThemeProvider as StyledThemeProvider } from 'styled-components';

import { ThemeContextType, ThemeType, isThemeType } from './types';
import { Nullable } from 'types/helpers';

import { darkTheme, lightTheme } from 'styles/themes';

const ThemeContext = createContext<Nullable<ThemeContextType>>(null);

export const useTheme = (): ThemeContextType => {
  const context = useContext(ThemeContext);

  if (!context) {
    throw new Error('useTheme must be used within a ThemeProvider');
  }

  return context;
};

type ThemeProviderProps = {
  children: ReactNode;
};

export const ThemeProvider = ({ children }: ThemeProviderProps) => {
  const [theme, setTheme] = useState<ThemeType>('dark');
  const isLightTheme = theme === 'light';

  const toggleTheme = () => {
    const updatedTheme = isLightTheme ? 'dark' : 'light';
    setTheme(updatedTheme);
    localStorage.setItem('theme', updatedTheme);
  };

  useEffect(() => {
    const savedTheme = localStorage.getItem('theme');
    const prefersLight = window.matchMedia && window.matchMedia('(prefers-color-scheme: light)').matches;
    const checkCurrentTheme = isThemeType(savedTheme);

    if (checkCurrentTheme) {
      setTheme(savedTheme);
    } else if (prefersLight) {
      setTheme('light');
    }
  }, []);

  return (
    <ThemeContext.Provider value={{ theme, toggleTheme }}>
      <StyledThemeProvider theme={isLightTheme ? lightTheme : darkTheme}>{children}</StyledThemeProvider>
    </ThemeContext.Provider>
  );
};
