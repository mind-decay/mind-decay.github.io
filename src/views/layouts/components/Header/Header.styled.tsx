import { Link } from 'react-router-dom';
import styled from 'styled-components';

import { screenBreakpoints } from 'styles/utils';

export const HeaderWrapper = styled.header`
  position: sticky;
  top: 0;
  z-index: 999;
  display: flex;
  justify-content: center;
  background-color: ${({ theme }) => theme.colors.subBG};
`;

export const HeaderNav = styled.nav`
  display: flex;
  align-items: center;
  justify-content: space-between;
  flex-grow: 1;
  max-width: ${screenBreakpoints.laptop}px;
  padding: 2.5rem 4rem;
  margin: 0 auto;

  @media screen and (max-width: ${screenBreakpoints.tablet}px) {
    padding: 2rem;
  }
`;

export const HeaderLogo = styled(Link)`
  display: flex;

  svg {
    width: 40px;
    height: auto;

    g {
      g {
        path {
          fill: ${({ theme }) => theme.colors.primary};
        }
      }
    }
  }
`;
