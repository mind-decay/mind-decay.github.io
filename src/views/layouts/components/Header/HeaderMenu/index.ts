export * from './HeaderMenu';
export * from './HeaderMenu.styled';
export * from './types';
export * from './routes';
