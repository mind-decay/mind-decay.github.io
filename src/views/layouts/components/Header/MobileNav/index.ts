export * from './MobileNav';
export * from './MobileNav.styled';
