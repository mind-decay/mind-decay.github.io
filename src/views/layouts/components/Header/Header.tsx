import { useState } from 'react';

import { HeaderLogo, HeaderNav, HeaderWrapper } from './Header.styled';
import { HeaderMenu, mainMenuOptions, menuRoutes } from './HeaderMenu';
import { MobileNav } from './MobileNav';

import Logo from 'assets/images/Logo.svg?react';

export const Header = () => {
  const [isMobileNavVisible, setIsMobileNavVisible] = useState(false);

  const handleMobileNavVisible = () => {
    setIsMobileNavVisible((prev) => !prev);
  };

  return (
    <>
      <HeaderWrapper>
        <HeaderNav>
          <HeaderLogo to={menuRoutes.Home} aria-label="Логотип при нажатии на который прокручивает до секции Интро">
            <Logo />
          </HeaderLogo>

          <HeaderMenu isVisible={isMobileNavVisible} onClick={handleMobileNavVisible} menuOptions={mainMenuOptions} />
        </HeaderNav>
      </HeaderWrapper>

      <MobileNav menuOptions={mainMenuOptions} isVisible={isMobileNavVisible} onClick={handleMobileNavVisible} />
    </>
  );
};
