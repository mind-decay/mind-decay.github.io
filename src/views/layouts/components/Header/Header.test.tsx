import { render } from '@testing-library/react';
import '@testing-library/jest-dom';

import { BrowserRouter } from 'react-router-dom';

import { ThemeProvider } from 'context/ThemeContext';

import { Header } from './Header';

test('renders header', () => {
  const { container } = render(
    <BrowserRouter>
      <ThemeProvider>
        <Header />
      </ThemeProvider>
    </BrowserRouter>,
  );

  expect(container).toMatchSnapshot();
});
