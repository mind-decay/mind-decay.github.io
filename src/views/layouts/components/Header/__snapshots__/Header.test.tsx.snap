// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`renders header 1`] = `
<div>
  <header
    class="sc-beySPh hkVEqb"
  >
    <nav
      class="sc-guDLey dbRwkn"
    >
      <a
        aria-label="Логотип при нажатии на который прокручивает до секции Интро"
        class="sc-dmyCSP klkhxa"
        href="/#home"
      >
        <svg
          data-name="Layer 2"
          id="Layer_2"
          viewBox="0 0 118.05 56.82"
          xmlns="http://www.w3.org/2000/svg"
        >
          <defs>
            <style>
              
      .cls-1 {
        fill: #fff;
        stroke-width: 0px;
      }
    
            </style>
          </defs>
          <g
            id="OBJECTS"
          >
            <g>
              <path
                class="cls-1"
                d="m70.38,51.48h-22.72c-1.99,0-3.6-1.61-3.6-3.6s1.61-3.6,3.6-3.6h22.72c1.99,0,3.6,1.61,3.6,3.6s-1.61,3.6-3.6,3.6Z"
              />
              <g>
                <path
                  class="cls-1"
                  d="m81.71,47.89c-.99,0-1.98-.41-2.69-1.21-1.32-1.49-1.18-3.76.31-5.08l11.45-10.15-11.45-10.15c-1.49-1.32-1.62-3.59-.31-5.08,1.32-1.49,3.59-1.62,5.08-.31l14.49,12.84c.77.68,1.21,1.66,1.21,2.69s-.44,2.01-1.21,2.69l-14.49,12.84c-.68.61-1.54.9-2.38.9Z"
                />
                <path
                  class="cls-1"
                  d="m36.34,47.89c-.85,0-1.7-.3-2.38-.9l-14.49-12.84c-.77-.68-1.21-1.66-1.21-2.69s.44-2.01,1.21-2.69l14.49-12.84c1.49-1.32,3.76-1.18,5.08.31,1.32,1.49,1.18,3.76-.31,5.08l-11.45,10.15,11.45,10.15c1.49,1.32,1.62,3.59.31,5.08-.71.8-1.7,1.21-2.69,1.21Z"
                />
              </g>
              <g>
                <path
                  class="cls-1"
                  d="m14.07,56.82c-.57-.35-14.07-8.8-14.07-28.41S13.5.35,14.07,0l4.16,6.83h.02c-.42.25-10.26,6.61-10.26,21.58s9.84,21.33,10.26,21.6h-.02s-4.16,6.81-4.16,6.81Z"
                />
                <path
                  class="cls-1"
                  d="m103.97,56.82l-4.16-6.83h-.02c.42-.25,10.26-6.61,10.26-21.58s-9.84-21.33-10.26-21.6h.02s4.16-6.81,4.16-6.81c.57.35,14.07,8.8,14.07,28.41s-13.5,28.06-14.07,28.41Z"
                />
              </g>
            </g>
          </g>
        </svg>
      </a>
      <div
        class="sc-eDLKkx gjhnoM"
      >
        <ul
          class="sc-jTQCzO etvuIx"
        >
          <li>
            <a
              aria-current="page"
              class="sc-gLLuof iYlPhz active"
              href="/#home"
            >
              Интро
            </a>
          </li>
          <li>
            <a
              aria-current="page"
              class="sc-gLLuof iYlPhz active"
              href="/#skills"
            >
              Навыки
            </a>
          </li>
          <li>
            <a
              aria-current="page"
              class="sc-gLLuof iYlPhz active"
              href="/#experience"
            >
              Опыт
            </a>
          </li>
          <li>
            <a
              aria-current="page"
              class="sc-gLLuof iYlPhz active"
              href="/#contact"
            >
              Связаться
            </a>
          </li>
        </ul>
        <button
          aria-label="Переключить тему"
          class="sc-hLQSwg kOBCup"
        >
          <svg
            fill="#ffffff"
            height="800px"
            id="Capa_1"
            stroke="#ffffff"
            viewBox="0 0 207.628 207.628"
            width="800px"
            xml:space="preserve"
            xmlns="http://www.w3.org/2000/svg"
            xmlns:xlink="http://www.w3.org/1999/xlink"
          >
            <g
              id="SVGRepo_bgCarrier"
              stroke-width="0"
            />
            <g
              id="SVGRepo_tracerCarrier"
              stroke-linecap="round"
              stroke-linejoin="round"
            />
            <g
              id="SVGRepo_iconCarrier"
            >
              <circle
                cx="103.814"
                cy="103.814"
                r="45.868"
              />
              <path
                d="M103.814,157.183c-29.427,0-53.368-23.941-53.368-53.368s23.941-53.368,53.368-53.368s53.368,23.941,53.368,53.368 S133.241,157.183,103.814,157.183z M103.814,65.446c-21.156,0-38.368,17.212-38.368,38.368s17.212,38.368,38.368,38.368 s38.368-17.212,38.368-38.368S124.97,65.446,103.814,65.446z"
              />
              <path
                d="M103.814,39.385c-4.142,0-7.5-3.358-7.5-7.5V7.5c0-4.142,3.358-7.5,7.5-7.5s7.5,3.358,7.5,7.5v24.385 C111.314,36.027,107.956,39.385,103.814,39.385z"
              />
              <path
                d="M103.814,207.628c-4.142,0-7.5-3.358-7.5-7.5v-24.385c0-4.142,3.358-7.5,7.5-7.5s7.5,3.358,7.5,7.5v24.385 C111.314,204.271,107.956,207.628,103.814,207.628z"
              />
              <path
                d="M200.128,111.314h-24.385c-4.142,0-7.5-3.358-7.5-7.5s3.358-7.5,7.5-7.5h24.385c4.142,0,7.5,3.358,7.5,7.5 S204.271,111.314,200.128,111.314z"
              />
              <path
                d="M31.885,111.314H7.5c-4.142,0-7.5-3.358-7.5-7.5s3.358-7.5,7.5-7.5h24.385c4.142,0,7.5,3.358,7.5,7.5 S36.027,111.314,31.885,111.314z"
              />
              <path
                d="M154.676,60.452c-1.919,0-3.839-0.732-5.303-2.197c-2.929-2.929-2.929-7.678,0-10.606l17.243-17.242 c2.929-2.929,7.678-2.93,10.606,0c2.929,2.929,2.929,7.678,0,10.606l-17.243,17.242C158.515,59.72,156.595,60.452,154.676,60.452z"
              />
              <path
                d="M35.709,179.419c-1.919,0-3.839-0.732-5.303-2.197c-2.929-2.929-2.929-7.678,0-10.606l17.243-17.243 c2.929-2.929,7.678-2.929,10.606,0c2.929,2.929,2.929,7.678,0,10.606l-17.243,17.243C39.548,178.687,37.629,179.419,35.709,179.419z "
              />
              <path
                d="M171.918,179.419c-1.919,0-3.839-0.732-5.303-2.197l-17.243-17.243c-2.929-2.929-2.929-7.678,0-10.606 c2.929-2.929,7.678-2.929,10.606,0l17.243,17.243c2.929,2.929,2.929,7.678,0,10.606 C175.757,178.687,173.838,179.419,171.918,179.419z"
              />
              <path
                d="M52.952,60.452c-1.919,0-3.839-0.732-5.303-2.197L30.406,41.013c-2.929-2.929-2.929-7.677,0-10.606 c2.929-2.929,7.678-2.93,10.606,0l17.243,17.242c2.929,2.929,2.929,7.677,0,10.606C56.791,59.72,54.872,60.452,52.952,60.452z"
              />
            </g>
          </svg>
        </button>
      </div>
    </nav>
  </header>
</div>
`;
